### PR TITLE
Implement cache mechanics with archive support

### PR DESCRIPTION
**Pull request by tormozz48**
Hello all. I have implemented following features:
- Optional cache model that can proxy all github calls
- Optional archive support
- Optional way for cache invalidation by cron job

@tadatuta @tavriaforever

### CODE REVIEW

**Comment by tavriaforever** [comment]
Here's what I don't understand about optimalnosti on/off the cache, well I have cut down in the configuration file and what?

**Comment by tormozz48** [comment]
Hello Mykola. At first I want to notice you that we already have moved our project to open source, so we must write all our comments in English language. 

If you disable cache option in config forum will work by the same way as it was earlier.

**Comment by tavriaforever** [comment]
OK, but why then the cache module built-in forum.js + variable names with the word cache

File: server/archive.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+module.exports = {
+
+    /**
+     * Initialize archive from local file
+     * @param options
+     * @returns {Archive}
+     */
+    init: function(options) {
+        console.log('Initialize archive start');
```
Comment (tadatuta) [inline server/archive.js]: Initialize archive start -> Initializing archive

File: server/archive.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
+var path = require('path'),
+    vowFs = require('vow-fs');
+
+var model;
```
Comment (tadatuta) [inline server/archive.js]: isn't it better to put initial object with empty `issues` and `comments` fields here and remove `onError`?

File: server/archive.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
+var path = require('path'),
```
Comment (tadatuta) [inline server/archive.js]: can't we treat this module as universal module to get posts from filesystem (not necessary archive)?
maybe someone will use just it without requests to github at all.

Comment (tadatuta) [inline server/archive.js]: shouldn't we implement it as a class to be able to inherit new objects and overload some methods?
e.g. to provide custom `getIssues` logic.

File: server/cache.js [hunk 4]
```diff
@@ -0,0 +1,442 @@
+var _ = require('lodash'),
+    vow = require('vow'),
+    CronJob = require('cron').CronJob,
+
+    github = require('./github'),
+    archive = require('./archive');
+
+var MAX_LIMIT = 100,
+    DEFAULT = {
+        PAGE: 1,
```
Comment (tadatuta) [inline server/cache.js]: I think we can use lower case for fields here. `DEFAULT` is kinda constant but it's value is just an ordinary object.

File: server/cache.js [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+        }
+    },
+    isCacheEnabled = false,
+    issues = [],
+    opts,
+    job;
+
+
+/**
+ * Loads All issues for configured github repository and returns them
```
Comment (tadatuta) [inline server/cache.js]: All -> all

File: server/cache.js [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+                pages;
+
+            //check for existed issues count for current repository
+            if(!count) {
+                issues = [];
+                return vow.resolve(issues);
+            }
+
+            //calculate number of pages
+            pages = ~~(count/MAX_LIMIT) + (count%MAX_LIMIT > 0 ? 1 : 0);
```
Comment (tadatuta) [inline server/cache.js]: `~~` looks too complicated to read

spaces around `%` needed

File: server/cache.js [hunk 7]
```diff
@@ -0,0 +1,442 @@
...
+            });
+        });
+}
+
+/**
+ * Initialize archive model and loads issues from it
+ * @param options - {Object} forum configuration object
+ * @returns {*}
+ */
+function loadArchiveIssues(options) {
```
Comment (tadatuta) [inline server/cache.js]: usage of archive should not be hardcoded

**Comment by tadatuta** [comment]
current implementation solves our own issues but is not abstract enough for universal solution.

I think `cache` module should be splited into two:
1. abstract interface to get some data (from gh or filesystem)
2. cache module itself which can be used by [1]

logic about dropping cache (by cron, events from gh, etc) should be separated and configurable.
and all the configuration about what is now called `archive` should be moved to `bem.info` repo.

but for sure you we can do it later.

File: server/archive.js (hunk 3 above)
Comment (tormozz48) [inline server/archive.js]: Can it be implemented in separate task later?

**Pull request merged**